includes/Core/Tags/Tag.php

```diff
@@ -0,0 +1,97 @@
+<?php
+/**
+ * Class Google\Site_Kit\Core\Tags\Tag
+ *
+ * @package   Google\Site_Kit\Core\Tags
+ * @copyright 2020 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Core\Tags;
+
+use Google\Site_Kit\Core\Guards\Guard_Interface;
+
+/**
+ * Base class for tags.
+ *
+ * @since n.e.x.t
+ * @access private
+ * @ignore
+ */
+abstract class Tag implements Tag_Interface {
+
+	/**
+	 * Module slug.
+	 *
+	 * @since n.e.x.t
+	 * @var string
+	 */
+	protected $slug;
+
+	/**
+	 * Tag ID.
+	 *
+	 * @since n.e.x.t
+	 * @var string
+	 */
+	protected $tag_id;
+
+	/**
+	 * Guards array.
+	 *
+	 * @since n.e.x.t
+	 * @var array
+	 */
+	protected $guards = array();
+
+	/**
+	 * Constructor.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $module_slug Module slug.
+	 * @param string $tag_id Tag ID.
+	 */
+	public function __construct( $module_slug, $tag_id ) {
+		$this->module_slug = $module_slug;
+		$this->tag_id      = $tag_id;
```

Comment:
I'm wary about this abstraction here. Since these classes are module-specific either way, I think it makes more sense to require a `Module_Settings` instance to be passed and handle the logic to get the ID from the settings in the `*_Tag` classes. There are other settings like `useSnippet` for example (which right now you're handling via the "guard" approach), but it could also be that a snippet requires two tags (e.g. we will need that for Google Analytics 4 support).

Handling the logic to get the respective settings and determine whether it should be rendered based on settings _inside_ of the `*_Tag` classes also removes more code from the parent class which I think makes sense. These module tag classes should encompass all logic related to rendering tags, which includes figuring out where the tag ID comes from etc. I think setting those things from the outside in the already bloated main module class makes this less intuitive to follow and maintain.